File: experiment_3/main.py
def find_all_lcs(X, Y, m, n, dp):
    if m == 0 or n == 0:
        return [[]]
    
    if X[m - 1] == Y[n - 1]:    # 如果X[m - 1] == Y[n - 1]，则X[m - 1]一定在LCS中，从左上角继续寻找
        lcs_list = find_all_lcs(X, Y, m - 1, n - 1, dp)
        for lcs in lcs_list:
            lcs.append(X[m - 1])
        return lcs_list
    else:
        lcs_list = []
        
        if dp[m - 1][n] >= dp[m][n - 1]:    # 如果dp[m - 1][n] >= dp[m][n - 1]，可以从上方继续寻找
            lcs_list.extend(find_all_lcs(X, Y, m - 1, n, dp))
        
        if dp[m][n - 1] >= dp[m - 1][n]:    # 如果dp[m][n - 1] >= dp[m - 1][n]，可以从左方继续寻找
            lcs_list.extend(find_all_lcs(X, Y, m, n - 1, dp))
        
        return lcs_list

def longest_common_subsequences(X, Y):
    m, n = len(X), len(Y)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if X[i - 1] == Y[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
    
    all_lcs = find_all_lcs(X, Y, m, n, dp)
    return dp[m][n], all_lcs

if __name__ == "__main__":
    # 读取输入数据
    T = int(input())
    for case in range(1, T + 1):
        m, n = map(int, input().split())
        X = input().split()
        Y = input().split()
        
        # 求最长公共子序列
        length, lcs_list = longest_common_subsequences(X, Y)
        
        # 输出结果
        print(f"Case {case}")
        print(f"{length} LCS(X,Y):")
        for lcs in lcs_list:
            print(" ".join(lcs))

File: experiment_3/main.cpp
#include <iostream>
#include <cstring>
#include <vector>
#include <algorithm>
using namespace std;

const int N = 60;

int m, n; // m = len(x), n = len(y)
int dp[N][N];
char x[N], y[N];
vector<vector<char>> lcs_list;

void find_all_lcs(int i, int j, vector<char> &lcs)
{
    if (i == 0 || j == 0) // 递归边界
    {
        reverse(lcs.begin(), lcs.end()); // 由于得到的是一个逆序的LCS，所以需要将其反转
        lcs_list.push_back(lcs);
        reverse(lcs.begin(), lcs.end()); // 恢复原来的顺序
        return;
    }

    if (x[i] == y[j]) // 如果x[i] == y[j]，则x[i]和y[j]一定在LCS中，在左上角的dp[i-1][j-1]处继续寻找
    {
        lcs.push_back(x[i]);
        find_all_lcs(i - 1, j - 1, lcs);
        lcs.pop_back();
    }
    else
    {
        if (dp[i - 1][j] == dp[i][j])
            find_all_lcs(i - 1, j, lcs);
        if (dp[i][j - 1] == dp[i][j])
            find_all_lcs(i, j - 1, lcs);
    }
}

void lcs()
{
    memset(dp, 0, sizeof(dp));

    for (int i = 1; i <= m; ++i)
    {
        for (int j = 1; j <= n; ++j)
        {
            if (x[i] == y[j])
                dp[i][j] = dp[i - 1][j - 1] + 1;
            else
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
        }
    }

    lcs_list.clear();
    vector<char> lcs;
    find_all_lcs(m, n, lcs);
}

int main()
{
    int T;
    cin >> T;

    for (int case_num = 1; case_num <= T; ++case_num)
    {
        cin >> m >> n;

        for (int i = 1; i <= m; ++i)
            cin >> x[i];

        for (int i = 1; i <= n; ++i)
            cin >> y[i];

        lcs();

        cout << endl
             << " "
             << "\t";
        for (int i = 1; i <= n; ++i)
            cout << y[i] << "\t";
        cout << endl;

        for (int i = 1; i <= m; ++i)
        {
            cout << x[i] << "\t";
            for (int j = 1; j <= n; ++j)
                cout << dp[i][j] << "\t";
            cout << endl;
        }
        cout << endl;

        for (auto &lcs : lcs_list)
        {
            for (auto &c : lcs)
                cout << c;
            cout << endl;
        }
    }

    system("pause");
    return 0;
}